cmd/storage-consumer/main.go

```diff
@@ -0,0 +1,525 @@
+	var partitionNum int64
+
+	str := `(\w+)\/(\w+)\/(\d+)\/(\d+\/)*`
+	switch dateSeparator {
+	case config.DateSeparatorNone.String():
+		str += `(\d{4})*`
+	case config.DateSeparatorYear.String():
+		str += `(\d{4})\/`
+	case config.DateSeparatorMonth.String():
+		str += `(\d{4}-\d{2})\/`
+	case config.DateSeparatorDay.String():
+		str += `(\d{4}-\d{2}-\d{2})\/`
+	}
+	str += `CDC(\d+).\w+`
+	pathRE, err := regexp.Compile(str)
+	if err != nil {
+		return 0, err
+	}
+
+	matches := pathRE.FindStringSubmatch(path)
+	if len(matches) != 7 {
+		return 0, fmt.Errorf("cannot match dml path pattern for %s", path)
+	}
+
+	version, err := strconv.ParseInt(matches[3], 10, 64)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(matches[4]) > 0 {
+		partitionNum, err = strconv.ParseInt(matches[4], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+	}
+	fileIdx, err := strconv.ParseUint(strings.TrimLeft(matches[6], "0"), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+
+	*d = dmlPathKey{
+		schemaPathKey: schemaPathKey{
+			schema:       matches[1],
+			table:        matches[2],
+			tableVersion: version,
+		},
+		partitionNum: partitionNum,
+		date:         matches[5],
+	}
+
+	return fileIdx, nil
+}
+
+// fileIndexRange defines a range of files. eg. CDC000002.csv ~ CDC000005.csv
+type fileIndexRange struct {
+	start uint64
+	end   uint64
+}
+
+type consumer struct {
+	sink            sink.Sink
+	replicationCfg  *config.ReplicaConfig
+	externalStorage storage.ExternalStorage
+	// tableIdxMap maintains a map of <dmlPathKey, max file index>
+	tableIdxMap map[dmlPathKey]uint64
+	// tableTsMap maintains a map of <dmlPathKey, max commit ts>
+	tableTsMap       map[dmlPathKey]uint64
+	tableIDGenerator *fakeTableIDGenerator
+}
+
+func newConsumer(ctx context.Context) (*consumer, error) {
+	replicaConfig := config.GetDefaultReplicaConfig()
+	err := util.StrictDecodeFile(configFile, "storage consumer", replicaConfig)
+	if err != nil {
+		log.Error("failed to decode config file", zap.Error(err))
+		return nil, err
+	}
+
+	err = replicaConfig.ValidateAndAdjust(upstreamURI)
+	if err != nil {
+		log.Error("failed to validate replica config", zap.Error(err))
+		return nil, err
+	}
+
+	if replicaConfig.Sink.Protocol != config.ProtocolCsv.String() {
+		return nil, fmt.Errorf("data encoded in protocol %s is not supported yet",
+			replicaConfig.Sink.Protocol)
+	}
+
+	bs, err := storage.ParseBackend(upstreamURIStr, nil)
+	if err != nil {
+		log.Error("failed to parse storage backend", zap.Error(err))
+		return nil, err
+	}
+
+	storage, err := storage.New(ctx, bs, &storage.ExternalStorageOptions{
+		SendCredentials: false,
+		S3Retryer:       rcommon.DefaultS3Retryer(),
+	})
+	if err != nil {
+		log.Error("failed to create external storage", zap.Error(err))
+		return nil, err
+	}
+
+	errCh := make(chan error, 1)
+	s, err := sink.New(ctx, model.DefaultChangeFeedID("storage-consumer"),
+		downstreamURIStr, config.GetDefaultReplicaConfig(), errCh)
```

Comment:
Is it should be downStream uri?

Reply:
At the consumer side, upstream will be cloud storage and downstream will be mysql/tidb, etc.